SU2_CFD/include/numerics/species/species_convection.hpp

```diff
@@ -0,0 +1,86 @@
+﻿/*!
+ * \file species_convection.hpp
+ * \brief Delarations of numerics classes for discretization of
+ *        convective fluxes in species problems.
+ * \author T. Kattmann
+ * \version 7.2.1 "Blackbird"
+ *
+ * SU2 Project Website: https://su2code.github.io
+ *
+ * The SU2 Project is maintained by the SU2 Foundation
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2021, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "../scalar/scalar_convection.hpp"
+
+/*!
+ * \class CUpwSca_Species
+ * \brief Class for doing a scalar upwind solver for the Spalar-Allmaras turbulence model equations.
+ * \ingroup ConvDiscr
+ */
+template <class FlowIndices>
+class CUpwSca_Species final : public CUpwScalar<FlowIndices> {
+ private:
+  using Base = CUpwScalar<FlowIndices>;
+  using Base::nVar;
+  using Base::nDim;
+  using Base::V_i;
+  using Base::V_j;
+  using Base::a0;
+  using Base::a1;
+  using Base::Flux;
+  using Base::Jacobian_i;
+  using Base::Jacobian_j;
+  using Base::ScalarVar_i;
+  using Base::ScalarVar_j;
+  using Base::idx;
+  /*!
+   * \brief Adds any extra variables to AD
+   */
+  void ExtraADPreaccIn() override {
+    AD::SetPreaccIn(V_i[idx.Density()]);
+    AD::SetPreaccIn(V_j[idx.Density()]);
+  };
+
+  /*!
+   * \brief SA specific steps in the ComputeResidual method
+   * \param[in] config - Definition of the particular problem.
+   */
+  void FinishResidualCalc(const CConfig* config) override {
+    for (auto iVar = 0u; iVar < nVar; iVar++) {
+      Flux[iVar] = a0 * V_i[idx.Density()] * ScalarVar_i[iVar] + a1 * V_j[idx.Density()] * ScalarVar_j[iVar];
+
+      /*--- Jacobians are taken wrt rho*Y not Y alone in the species solver. ---*/
+      /*--- Off-diagonal entries are zero. ---*/
+      Jacobian_i[iVar][iVar] = a0;
+      Jacobian_j[iVar][iVar] = a1;
```

Comment:
Try initializing the Jacobian buffer in CUpwScalar and AvgGrad with all zeros

Reply:
I should have that... but it's one of those things gcc 4.8.5 messes up sometimes, and then we have people complaining :shrug:

Reply:
4b6ae60 looks like that's it ... valgrind is silenced and my local tests are positive. For some reason I was fully focused on the Flux and overlooked the JacobianBuffer although I stared at the diffusion files for quite a bit 🤦  

Note to self: Post debugging-findings earlier next time. 

Thanks a lot man 🙇‍♀️

Reply:
I wont complain ... i was also bound to 5.3.0 until very recently. I will put the zero-ing it explicitly in the constructor then

Reply:
I can see if it compiles with 4.8.5, we have other examples, sometimes it works, sometimes it doesn't

Reply:
Yes sir, early findings is the way :)

Reply:
Now explicitly set to zero in ebf853e